invenio_vocabularies/contrib/names/datastreams.py

```diff
@@ -67,42 +68,55 @@ def _process_lambda_file(self, fileobj):
         if self.since:
             time_shift = self.since
         last_sync = arrow.now() - timedelta(**time_shift)
+        try:
+            file_content = fileobj.read().decode("utf-8")
 
-        file_content = fileobj.read().decode("utf-8")
-
-        csv_reader = csv.DictReader(file_content.splitlines())
+            csv_reader = csv.DictReader(file_content.splitlines())
 
-        for row in csv_reader:  # Skip the header line
-            orcid = row["orcid"]
+            for row in csv_reader:  # Skip the header line
+                orcid = row["orcid"]
 
-            # Lambda file is ordered by last modified date
-            last_modified_str = row["last_modified"]
-            try:
-                last_modified_date = arrow.get(last_modified_str, date_format)
-            except arrow.parser.ParserError:
-                last_modified_date = arrow.get(last_modified_str, date_format_no_millis)
+                # Lambda file is ordered by last modified date
+                last_modified_str = row["last_modified"]
+                try:
+                    last_modified_date = arrow.get(last_modified_str, date_format)
+                except arrow.parser.ParserError:
+                    last_modified_date = arrow.get(
+                        last_modified_str, date_format_no_millis
+                    )
 
-            if last_modified_date < last_sync:
-                break
-            yield orcid
+                if last_modified_date < last_sync:
+                    break
+                yield orcid
+        finally:
+            fileobj.close()
 
     def _iter(self, orcids):
         """Iterates over the ORCiD records yielding each one."""
         with ThreadPoolExecutor(
             max_workers=current_app.config["VOCABULARIES_ORCID_SYNC_MAX_WORKERS"]
         ) as executor:
-            futures = [
+            # futures is a dictionary where the key is the Future object
+            futures = {
                 executor.submit(
                     self._fetch_orcid_data,
                     orcid,
                     current_app.config["VOCABULARIES_ORCID_SUMMARIES_BUCKET"],
-                )
+                ): orcid
                 for orcid in orcids
-            ]
+            }
+
             for future in as_completed(futures):
-                result = future.result()
-                if result is not None:
-                    yield result
+                try:
+                    result = future.result()
+                    if result:
+                        yield result
+                finally:
+                    # Explicitly release memory, as we don't need the future anymore.
+                    # This is mostly required because as long as we keep a reference to the future
+                    # (in the above futures dict), the garbage collector won't collect it
+                    # and it will keep the memory allocated.
+                    futures.pop(future, None)
```

Comment:
minor: isn't it a bit dangerous to modify the `futures` dictionary while iterating over it? The alternative way would be to reverse the key/value above (i.e. `orcid: executor.submit(...)` inside the dict comprehension), and then iterate over `list(future.keys())` or similar.